File: leetcode/algorithms/121_best_time_to_buy_and_sell_stock/main.c
#include <limits.h>

/**
 * Greedy
 * 
 * Complexities:
 *   N - `pricesSize`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(1)
 */
int maxProfit(int* prices, int pricesSize) {
    int profit = 0;
    int minPrice = 100000;

    for (int i = 0; i < pricesSize; i++) {
        if (prices[i] < minPrice) {
            minPrice = prices[i];
        }

        int diff = prices[i] - minPrice;
        if (diff > profit) {
            profit = diff;
        }
    }

    return profit;
}


// Solution
/**
 * Greedy
 * 
 * Complexities:
 *   N - `pricesSize`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(1)
 */
int maxProfit(int* prices, int pricesSize) {
    if (pricesSize < 2) {
        return 0;
    }

    int min_price = INT_MAX;
    int max_profit = 0;

    for (int i = 0; i < pricesSize; i++) {
        if (prices[i] < min_price) {
            min_price = prices[i];
        } else if (prices[i] - min_price > max_profit) {
            max_profit = prices[i] - min_price;
        }
    }

    return max_profit;
}

File: leetcode/algorithms/121_best_time_to_buy_and_sell_stock/main.java
class BestTimeToBuyAndSellStock {
  /**
   * Greedy
   *
   * Complexities:
   *   N - Size of `prices`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(1)
   */
  public int maxProfit(int[] prices) {
    int minPrice = 100000;
    int maxProfit = 0;

    for (int price : prices) {
      if (price < minPrice) {
        minPrice = price;
      }

      int profit = price - minPrice;
      if (profit > maxProfit) {
        maxProfit = profit;
      }
    }

    return maxProfit;
  }


  // Solution
  /**
   * Greedy
   *
   * Complexities:
   *   N - Size of `prices`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(1)
   */
  public int solution(int[] prices) {
    if (prices == null || prices.length <= 1) {
      return 0;
    }

    int minPrice = Integer.MAX_VALUE;
    int maxProfit = 0;

    for (int price : prices) {
      if (price < minPrice) {
        minPrice = price;
      } else if (price - minPrice > maxProfit) {
        maxProfit = price - minPrice;
      }
    }

    return maxProfit;
  }
}
